File: clients/android/app/src/main/kotlin/com/lextures/android/core/lms/PathsLogic.kt
package com.lextures.android.core.lms

object PathsLogic {
    val recommendationSurfaces = listOf("continue", "review", "strengthen", "challenge")

    fun sortedCourses(courses: List<PathCourseProgress>): List<PathCourseProgress> =
        courses.sortedBy { it.position }

    fun nextCourse(progress: PathProgress): PathCourseProgress? =
        sortedCourses(progress.courses).firstOrNull { !it.isCompleted }

    fun isLocked(course: PathCourseProgress): Boolean =
        !course.isCompleted && !course.isRecommended

    fun isPaid(bundlePriceCents: Int?): Boolean =
        bundlePriceCents != null && bundlePriceCents > 0

    fun catalogWebPath(slug: String): String = "/paths/$slug"

    fun formatPrice(cents: Int, currency: String = "USD"): String {
        val formatter = java.text.NumberFormat.getCurrencyInstance()
        formatter.currency = java.util.Currency.getInstance(currency)
        return formatter.format(cents / 100.0)
    }

    fun formatDuration(minutes: Int): String = when {
        minutes < 60 -> "$minutes min"
        minutes % 60 == 0 -> "${minutes / 60} hr"
        else -> "${minutes / 60} hr ${minutes % 60} min"
    }

    data class MergedRecommendations(
        val primary: LearnerRecommendationItem?,
        val chips: List<LearnerRecommendationItem>,
        val degraded: Boolean,
    )

    fun mergeRecommendations(responses: List<LearnerRecommendationsResponse>): MergedRecommendations {
        val merged = responses.flatMap { it.recommendations }.sortedByDescending { it.score }
        return MergedRecommendations(
            primary = merged.firstOrNull(),
            chips = merged.drop(1).take(3),
            degraded = responses.any { it.degraded == true },
        )
    }

    fun structureItemKind(itemType: String): String? = when (itemType) {
        "quiz" -> "quiz"
        "content_page" -> "content_page"
        "assignment" -> "assignment"
        "external_link" -> "external_link"
        else -> null
    }

    fun surfaceLabelRes(surface: String): Int = when (surface) {
        "continue" -> com.lextures.android.R.string.mobile_paths_surface_continue
        "strengthen" -> com.lextures.android.R.string.mobile_paths_surface_strengthen
        "challenge" -> com.lextures.android.R.string.mobile_paths_surface_challenge
        "review" -> com.lextures.android.R.string.mobile_paths_surface_review
        else -> com.lextures.android.R.string.mobile_paths_surface_other
    }

    fun structureItem(
        recommendation: LearnerRecommendationItem,
        items: List<CourseStructureItem>,
    ): CourseStructureItem? {
        items.firstOrNull { it.id == recommendation.itemId }?.let { return it }
        val kind = structureItemKind(recommendation.itemType) ?: return null
        return CourseStructureItem(
            id = recommendation.itemId,
            sortOrder = 0,
            kind = kind,
            title = recommendation.title,
            parentId = null,
            published = true,
            dueAt = null,
            pointsWorth = null,
            pointsPossible = null,
        )
    }
}

File: clients/ios/Lextures/Core/LMS/PathsLogic.swift
import Foundation

enum PathsLogic {
    static let recommendationSurfaces = ["continue", "review", "strengthen", "challenge"]

    static func sortedCourses(_ courses: [PathCourseProgress]) -> [PathCourseProgress] {
        courses.sorted { $0.position < $1.position }
    }

    static func nextCourse(in progress: PathProgress) -> PathCourseProgress? {
        sortedCourses(progress.courses).first { !$0.isCompleted }
    }

    static func isLocked(_ course: PathCourseProgress) -> Bool {
        !course.isCompleted && !course.isRecommended
    }

    static func isPaid(bundlePriceCents: Int?) -> Bool {
        guard let bundlePriceCents else { return false }
        return bundlePriceCents > 0
    }

    static func catalogWebPath(slug: String) -> String {
        "/paths/\(slug)"
    }

    static func formatPrice(cents: Int, currency: String = "USD") -> String {
        let major = CurrencyExponent.minorUnitsToMajorUnits(cents, currency: currency)
        let formatter = NumberFormatter()
        formatter.numberStyle = .currency
        formatter.currencyCode = currency
        return formatter.string(from: NSNumber(value: major)) ?? "\(cents)"
    }

    static func formatDuration(minutes: Int) -> String {
        if minutes < 60 { return L.format("mobile.paths.duration.minutes", minutes) }
        let hours = minutes / 60
        let rem = minutes % 60
        if rem == 0 { return L.format("mobile.paths.duration.hours", hours) }
        return L.format("mobile.paths.duration.hoursMinutes", hours, rem)
    }

    static func mergeRecommendations(
        _ responses: [LearnerRecommendationsResponse]
    ) -> (primary: LearnerRecommendationItem?, chips: [LearnerRecommendationItem], degraded: Bool) {
        var merged: [LearnerRecommendationItem] = []
        var degraded = false
        for response in responses {
            merged.append(contentsOf: response.recommendations)
            if response.degraded == true { degraded = true }
        }
        merged.sort { $0.score > $1.score }
        return (merged.first, Array(merged.dropFirst().prefix(3)), degraded)
    }

    static func structureItemKind(for itemType: String) -> String? {
        switch itemType {
        case "quiz": return "quiz"
        case "content_page": return "content_page"
        case "assignment": return "assignment"
        case "external_link": return "external_link"
        default: return nil
        }
    }

    static func surfaceLabelKey(_ surface: String) -> String {
        switch surface {
        case "continue": return "mobile.paths.surface.continue"
        case "strengthen": return "mobile.paths.surface.strengthen"
        case "challenge": return "mobile.paths.surface.challenge"
        case "review": return "mobile.paths.surface.review"
        default: return "mobile.paths.surface.other"
        }
    }

    static func structureItem(
        for recommendation: LearnerRecommendationItem,
        in items: [CourseStructureItem]
    ) -> CourseStructureItem? {
        if let existing = items.first(where: { $0.id == recommendation.itemId }) {
            return existing
        }
        guard let kind = structureItemKind(for: recommendation.itemType) else { return nil }
        return CourseStructureItem(
            id: recommendation.itemId,
            sortOrder: 0,
            kind: kind,
            title: recommendation.title,
            parentId: nil,
            published: true,
            dueAt: nil,
            pointsWorth: nil,
            pointsPossible: nil
        )
    }
}